crypto/ct/ct_lib.c

```diff
@@ -0,0 +1,301 @@
+    if (logidlen != SCT_V1_HASHLEN) {
+        CTerr(CT_F_SCT_SET0_LOGID, CT_R_INVALID_LOGID_LENGTH);
+        return 0;
+    }
+    if (sct->logid)
+        OPENSSL_free(sct->logid);
+    sct->logid = logid;
+    sct->logidlen = logidlen;
+    return 1;
+}
+
+int SCT_set_timestamp(SCT *sct, uint64_t timestamp)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SET_TIMESTAMP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    sct->timestamp = timestamp;
+    return 1;
+}
+
+int SCT_set_signature_nid(SCT *sct, int nid)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SET_SIGNATURE_NID, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (nid == NID_sha256WithRSAEncryption)
+        sct->sig_alg = TLSEXT_signature_rsa;
+    else if (nid == NID_ecdsa_with_SHA256)
+        sct->sig_alg = TLSEXT_signature_ecdsa;
+    else
+        return 0;
+    sct->hash_alg = TLSEXT_hash_sha256;
+    return 1;
+}
+
+int SCT_set0_extensions(SCT *sct, unsigned char *ext, size_t extlen)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SET0_EXTENSIONS, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (sct->ext)
+        OPENSSL_free(sct->ext);
+    sct->ext = ext;
+    sct->extlen = extlen;
+    return 1;
+}
+
+int SCT_set0_signature(SCT *sct, unsigned char *sig, size_t siglen)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SET0_SIGNATURE, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (sct->sig)
+        OPENSSL_free(sct->sig);
+    sct->sig = sig;
+    sct->siglen = siglen;
+    return 1;
+}
+
+int SCT_get_version(const SCT *sct, sct_version_t *version)
+{
+    if (sct == NULL || version == NULL) {
+        CTerr(CT_F_SCT_GET_VERSION, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *version = sct->version;
+    return 1;
+}
+
+int SCT_get_log_entry_type(SCT *sct, log_entry_type_t *entry_type)
+{
+    if (sct == NULL || entry_type == NULL) {
+        CTerr(CT_F_SCT_GET_LOG_ENTRY_TYPE, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *entry_type = sct->entry_type;
+    return 1;
+}
+
+int SCT_get0_logid(const SCT *sct, unsigned char **logid, size_t *logidlen)
+{
+    if (sct == NULL || logid == NULL || logidlen == NULL) {
+        CTerr(CT_F_SCT_GET0_LOGID, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *logid = sct->logid;
+    *logidlen = sct->logidlen;
+    return 1;
+}
+
+int SCT_get_timestamp(const SCT *sct, uint64_t *timestamp)
+{
+    if (sct == NULL || timestamp == NULL) {
+        CTerr(CT_F_SCT_GET_TIMESTAMP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *timestamp = sct->timestamp;
+    return 1;
+}
+
+int SCT_get_signature_nid(const SCT *sct)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_GET_SIGNATURE_NID, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
+    if (sct->version != SCT_V1) {
+        CTerr(CT_F_SCT_GET_SIGNATURE_NID, CT_R_SCT_UNSUPPORTED_VERSION);
+        return -1;
```

Comment:
I don't particularly like returning -1 here, even if it's sort of an undocumented contract that NIDs are positive. Maybe just return NID_undef.